sklearn/tests/test_docstring_parameters.py

```diff
@@ -0,0 +1,173 @@
+"""
+Test mismatches between docstring and method/function signature.
+"""
+
+# Authors: Joel Nothman <joel.nothman@gmail.com>
+# License: BSD 3 clause
+
+
+# TODO inspect for Cython (see sagenb.misc.sageinspect)
+from __future__ import print_function
+
+import os
+import inspect
+import imp
+import sys
+import sklearn
+import warnings
+from pkgutil import walk_packages
+from nose import SkipTest
+
+
+scikit_learn_path = os.getcwd().replace("/sklearn/tests", "")
+docscrape_path = 'doc/sphinxext/numpy_ext/docscrape.py'
+docscrape = imp.load_source('docscrape',
+                            os.path.join(scikit_learn_path, docscrape_path))
+
+# To capture warnings related to invalid section names in docs
+warnings.simplefilter('always')
+
+
+def get_name(func):
+    parts = []
+    module = inspect.getmodule(func)
+    if module:
+        parts.append(module.__name__)
+    if hasattr(func, 'im_class'):
+        parts.append(func.im_class.__name__)
+    parts.append(func.__name__)
+    return '.'.join(parts)
+
+
+def check_parameters_match(func, doc=None):
+    mismatches = 0
+
+    name = get_name(func)
+    if not name.startswith('sklearn.'):
+        return 0
+    if inspect.isdatadescriptor(func):
+        return 0
+
+    # Ignore abstract methods.
+    source = inspect.getsource(func).strip()
+    if source.endswith('\"\"\"') or source.endswith('\'\'\'') or \
+            source.endswith('pass'):
+        return 0
+
+    try:
+        args, varargs, varkw, defaults = inspect.getargspec(func)
+    except TypeError:
+        return 0
+
+    if 'self' in args:
+        args.remove('self')
+
+    if doc is None:
+        with warnings.catch_warnings(record=True) as w:
+            try:
+                doc = docscrape.FunctionDoc(func)
+            except ValueError:
+                print('Failed to parse doc of %r' % func, file=sys.stderr)
+                return 0
+            msg = ('Warning was raised when attempting to scrape the doc '
+                   'of the class with name %s.\n'
+                   'The warning was: %s ')
+            if len(w) != 0:
+                print(msg % (name, w[0].message))
+                mismatches += 1
+
+    # check set
+    param_names = [name for name, _, _ in doc['Parameters']]
+    # clean up some docscrape output:
+    param_names = [name.split(':')[0].strip('` ') for name in param_names]
+    param_names = [name for name in param_names if '*' not in name]
+
+    try:
+        args_set = set(args)
+    except TypeError:
+        # TODO: handle arg tuples
+        return 0
+    extra_params = set(param_names) - args_set
+    if extra_params and not varkw:
+        print(get_name(func), 'in doc', sorted(extra_params))
+        mismatches += 1
+
+    if defaults:
+        none_defaults = [arg for arg, default in zip(args[-len(defaults):],
```

Comment:
This exclude parameters in signature which are initialized to None by default... I think this was done so that the optional y need not be documented...